packages/dom-adapters/src/CaretAdapter/index.ts

```diff
@@ -0,0 +1,194 @@
+import type { Caret, EditorJSModel, CaretManagerEvents, Index } from '@editorjs/model';
+import { getAbsoluteRangeOffset, getBoundaryPointByAbsoluteOffset, useSelectionChange } from '../utils/index.js';
+import type { TextRange } from '@editorjs/model';
+import { EventType, IndexBuilder } from '@editorjs/model';
+import { isNativeInput } from '@editorjs/dom';
+
+/**
+ * Caret adapter watches selection change and saves it to the model
+ *
+ * On model update, it updates the selection in the DOM
+ */
+export class CaretAdapter extends EventTarget {
+  /**
+   * Editor.js DOM container
+   *
+   * @private
+   */
+  #container: HTMLElement;
+
+  /**
+   * Editor.js model
+   *
+   * @private
+   */
+  #model: EditorJSModel;
+
+  /**
+   * Map of inputs
+   *
+   * @private
+   */
+  #inputs = new Map<string, HTMLElement>();
+
+  /**
+   * Current user's caret
+   *
+   * @private
+   */
+  #userCaret: Caret;
+
+  /**
+   * @class
+   * @param container - Editor.js DOM container
+   * @param model - Editor.js model
+   */
+  constructor(container: HTMLElement, model: EditorJSModel) {
+    super();
+
+    this.#model = model;
+    this.#container = container;
+    this.#userCaret = this.#model.createCaret();
+
+    const { on } = useSelectionChange();
+
+    on(container, (selection) => this.#onSelectionChange(selection), this);
+
+    this.#model.addEventListener(EventType.CaretManagerUpdated, (event) => this.#onModelUpdate(event));
+  }
+
+  /**
+   * Adds input to the caret adapter
+   *
+   * @param input - input element
+   * @param index - index of the input in the model tree
+   */
+  public attachInput(input: HTMLElement, index: Index): void {
+    this.#inputs.set(index.serialize(), input);
+  }
+
+  /**
+   * Updates current user's caret index
+   *
+   * @param index - new caret index
+   */
+  public updateIndex(index: Index): void {
+    this.#userCaret.update(index);
+  }
+
+  /**
+   * Selection change handler
+   *
+   * @param selection - new document selection
+   */
+  #onSelectionChange(selection: Selection | null): void {
+    if (!selection) {
+      return;
+    }
+
+    const activeElement = document.activeElement;
+
+    for (const [index, input] of this.#inputs) {
+      if (input !== activeElement) {
+        continue;
+      }
+
+      if (isNativeInput(input) === true) {
+        const textRange = [
+          (input as HTMLInputElement | HTMLTextAreaElement).selectionStart,
+          (input as HTMLInputElement | HTMLTextAreaElement).selectionEnd,
+        ] as TextRange;
+
+        const builder = new IndexBuilder();
+
+        builder.from(index).addTextRange(textRange);
+
+        this.updateIndex(builder.build());
+
+        /**
+         * For now we handle only first found input
+         */
+        break;
+      }
+
```

Comment:
the page freezes and dies when I focus contenteditable input in Safari

Reply:
Fixed